File: 01189. Maximum Number of Balloons/solution.go
/**
 * Problem: 1189. Maximum Number of Balloons
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func maxNumberOfBalloons(text string) int {
	// Create a map to count occurrences of each rune (character) in the input string
	m := make(map[rune]int)

	// Iterate over each character in the input string and increment its count in the map
	for _, r := range text {
		m[r]++
	}
	
	// Calculate the maximum number of "balloon" words that can be formed.
	// 'l' and 'o' appear twice in "balloon", so their counts are divided by 2.
	// The minimum count among 'b', 'a', 'l'/2, 'o'/2, and 'n' determines the answer.
	return min(m['b'], m['a'], m['l']/2, m['o']/2, m['n'])
}

File: 01189. Maximum Number of Balloons/solution.cpp
/**
 * Problem: 1189. Maximum Number of Balloons
 *
 * Difficulty: Easy
 *
 * Language: C++
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

class Solution
{
public:
  int maxNumberOfBalloons(string text)
  {
    // Create a hash map to store frequency of each character
    unordered_map<char, int> charFrequency;

    // Count occurrences of each character in the input string
    for (char character : text)
      charFrequency[character]++;

    // Calculate maximum number of "balloon" words that can be formed
    // 'b', 'a', 'n' appear once, 'l' and 'o' appear twice in "balloon"
    return min({charFrequency['b'], charFrequency['a'],
                charFrequency['l'] / 2, charFrequency['o'] / 2,
                charFrequency['n']});
  }
};
